server/config/validate-config-keys.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * External dependencies
+ */
+const chalk = require( 'chalk' );
+const fs = require( 'fs' );
+const path = require( 'path' );
+
+/** @type {String} path to configuration file directory */
+const configRoot = path.resolve( __dirname, '../../', 'config' );
+
+/**
+ * @type {RegExp} matches possible config files
+ *
+ *  - doesn't start with an underscore
+ *  - ends in `.json`
+ */
+const environmentConfigPattern = /^(?!_)[^.]+\.json$/;
+
+/**
+ * Reads a config file given its basename
+ *
+ * @param {String} filename basename of config file to read, e.g. 'development.json'
+ * @returns {String} contents of file
+ */
+const readConfigFile = filename => fs.readFileSync( path.join( configRoot, filename ), { encoding: 'utf8' } );
+
+/**
+ * Reads and parses the data from a
+ * config file given its basename
+ *
+ * @throws SyntaxError if contents of config file not valid JSON
+ * @param {String} filename basename of config file to read, e.g. 'development.json'
+ * @returns {*} parsed data from config file contents
+ */
+const parseConfig = filename => JSON.parse( readConfigFile( filename ) );
+
+/** @type {Array} list of [ filename, config data keys ] configuration pairs */
+const environmentKeys = fs
+	.readdirSync( configRoot, { encoding: 'utf8' } )
+	.filter( filename => environmentConfigPattern.test( path.basename( filename ) ) )
+	.filter( filename => ! ( /secrets/g ).test( filename ) )
```

Comment:
I got a bit confused by this filter section at first.
the environmentConfigPattern regex has a section in it meant just to `filter` out _shared.json, whereas you specifically call out these two files here later.  And also if the regexes are meant to rule out single files, why not match against exact strings which is a bit simpler?

Personally I think it might be clearest to make the regex grab all json files then filter the filenames out considering there are only 3. 

`environmentConfigPattern` --> `jsonPattern`

then the filter section becomes 
```js
  .filter( filename => filename !== '_shared.json' )
  .filter( filename => filename !== 'secrets.json' )
  .filter( filename => filename !== 'client.json ' )
```

feel free to disagree, this shouldn't block merging this 👍